File: epi_judge_java_solutions/epi/Knapsack.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiUserType;
import epi.test_framework.GenericTest;

import java.util.Arrays;
import java.util.List;

public class Knapsack {
  @EpiUserType(ctorParams = {Integer.class, Integer.class})

  public static class Item {
    public Integer weight;
    public Integer value;

    public Item(Integer weight, Integer value) {
      this.weight = weight;
      this.value = value;
    }
  }

  @EpiTest(testDataFile = "knapsack.tsv")

  public static int optimumSubjectToCapacity(List<Item> items, int capacity) {

    // V[i][j] holds the optimum value when we choose from
    // items.subList(0, i + 1) and have a capacity of j.
    int[][] V = new int[items.size()][capacity + 1];
    for (int[] v : V) {
      Arrays.fill(v, -1);
    }
    return optimumSubjectToItemAndCapacity(items, items.size() - 1, capacity,
                                           V);
  }

  // Returns the optimum value when we choose from items.subList(0, k + 1) and
  // have a capacity of availableCapacity.
  private static int optimumSubjectToItemAndCapacity(List<Item> items, int k,
                                                     int availableCapacity,
                                                     int[][] V) {
    if (k < 0) {
      // No items can be chosen.
      return 0;
    }

    if (V[k][availableCapacity] == -1) {
      int withoutCurrItem =
          optimumSubjectToItemAndCapacity(items, k - 1, availableCapacity, V);
      int withCurrItem =
          availableCapacity < items.get(k).weight
              ? 0
              : items.get(k).value +
                    optimumSubjectToItemAndCapacity(
                        items, k - 1, availableCapacity - items.get(k).weight,
                        V);
      V[k][availableCapacity] = Math.max(withoutCurrItem, withCurrItem);
    }
    return V[k][availableCapacity];
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "Knapsack.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python_solutions/knapsack.py
import collections
import functools
from typing import List

from test_framework import generic_test
from test_framework.test_utils import enable_executor_hook

Item = collections.namedtuple('Item', ('weight', 'value'))


def optimum_subject_to_capacity(items: List[Item], capacity: int) -> int:

    # Returns the optimum value when we choose from items[:k + 1] and have a
    # capacity of available_capacity.
    @functools.lru_cache(None)
    def optimum_subject_to_item_and_capacity(k, available_capacity):
        if k < 0:
            # No items can be chosen.
            return 0

        without_curr_item = optimum_subject_to_item_and_capacity(
            k - 1, available_capacity)
        with_curr_item = (0 if available_capacity < items[k].weight else
                          (items[k].value +
                           optimum_subject_to_item_and_capacity(
                               k - 1, available_capacity - items[k].weight)))
        return max(without_curr_item, with_curr_item)

    return optimum_subject_to_item_and_capacity(len(items) - 1, capacity)


@enable_executor_hook
def optimum_subject_to_capacity_wrapper(executor, items, capacity):
    items = [Item(*i) for i in items]
    return executor.run(
        functools.partial(optimum_subject_to_capacity, items, capacity))


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('knapsack.py', 'knapsack.tsv',
                                       optimum_subject_to_capacity_wrapper))
